docs/11-decorators.md

```diff
@@ -42,6 +37,33 @@ ActiveAdmin.register Post do
 end
 ```
 
+If you wish to use a PORO decorator instead, all you need to do is
+define the `.decorate` method on the class. With that change the `PostDecorator`
+from above would look like this:
+
+```ruby
+# app/decorators/post_decorator.rb
+class PostDecorator
+  attr_reader :post
+  delegate_missing_to :post
+
+  def initialize(post)
+    @post = post
+  end
+
+  def self.decorate(record_or_collection, *_)
+    if record_or_collection.is_a?(Enumerable)
+      record_or_collection.map { |individual_record| new(individual_record) }
+    else
+      new(record_or_collection)
+    end
+  end
+end
```

Comment:
Something I don't love about this is that each decorator needs to add this code, which is as you point out later very generic. I feel ActiveAdmin should just work without having to add this. So I took a stab at it by copying the basic functionality from draper's collection presenter. I came up with this:

```diff
--- a/lib/active_admin/resource_controller/decorators.rb
+++ b/lib/active_admin/resource_controller/decorators.rb
@@ -48,18 +48,38 @@ module ActiveAdmin
         end
       end
 
+      class CollectionDecorator
+        # @return the collection being decorated.
+        attr_reader :object
+
+        # @return [Class] the decorator class used to decorate each item, as set by {#initialize}.
+        attr_reader :decorator_class
+
+        array_methods = Array.instance_methods - Object.instance_methods
+        delegate :==, :as_json, *array_methods, to: :decorated_collection
+
+        def initialize(object, with:)
+          @object = object
+          @decorator_class = with
+        end
+
+        class << self
+          alias_method :decorate, :new
+        end
+
+        def decorated_collection
+          @decorated_collection ||= object.map { |item| decorator_class.new(item) }
+        end
+      end
+
       class Wrapper
         @cache = {}
 
         def self.wrap(decorator)
-          collection_decorator = find_draper_collection_decorator(decorator)
+          collection_decorator = find_draper_collection_decorator(decorator) || CollectionDecorator
 
-          if collection_decorator
-            name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
-            @cache[name] ||= wrap! collection_decorator, name
-          else
-            decorator
-          end
+          name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
+          @cache[name] ||= wrap! collection_decorator, name
         end
 
         private
```

With that in place, I believe we don't require any other changes whatsoever to support PORO decorators, and people can just plugin their own decorators, regardless of whether they are draper decorators or other types.

What do you think?

Reply:
I like the idea of allowing people to use decorators without needing to add any custom code. My only concern would be that we are forcing the use of `#new` with just one parameter and not making it very flexible. So if I wanted to initialize my presenter from many different sources, I'd like to use something like a Factory Method like:

```ruby
class PostPresenter
  def self.wrap(post, *_)
    self.for_admin(post)
  end

  def self.for_admin(post)
    # ...
  end
end
```

But after writing that code I guess what I'd rather do is an `AdminPostPresenter` class instead.

In short, I agree with the change. I'll implement it and change the docs accordingly if that's alright.

Reply:
Right.

Maybe instead of explicitly checking for `draper`, we could use `decorator.collection_decorator_class` if it's defined. As long as we document that expected interface `.decorate(<collection>, with: <decorator_class>)`, people should be able to override that to fit their needs.

Reply:
Yes, go ahead!

Oh, I forgot I also needed to add `, to: :object` add the end of the `Wrapper.wrap` delegations so that it uses the standard Rails delegation instead of the `draper` shortcut.

Reply:
Yeah, no worries I got that 😄 . BTW with this change, we wouldn't need to delegate or call those methods at all. Since wrapping would be done on the render level all we need are the array methods to iterate. Though I think small changes are better

Reply:
My idea is to keep the current wrapping at the same level, and ideally just touch code paths that don't affect the `draper` integration at all. So ideally revert all changes at `lib/active_admin/resource_controller.rb`, `lib/active_admin/resource_controller/data_access.rb`, `lib/active_admin/views/pages/index.rb`, and restore the removed specs.

Maybe delaying the decoration is a good thing, but I'd like to minimize the chance of regressions because of the addition of this feature.

Reply:
Agreed, I made the proposed changes. Let me know if you have any other comments, happy to implement them.

Reply:
Thanks for your work! I'll do another review pass tomorrow, but should be essentially ready for my side :)